### PR TITLE
Clarify the 'promises' exercise instructions

### LINKED ISSUE #1331: Improve Practice Exercise "Promises"
I was trying to help a student but the exercise makes no sense to me.  If I implement just a TINY implementation:

```
export const promisify = (fn) => {
  return (v) => {
    return new Promise(() => {
    });
  };
};
```

All the tests start turning green, despite the fact that I'm not even calling `fn`, not rejecting or resolving anything... if I can't trust the test suite to guide me then I'm not sure how any student is supposed to get thru this complex exercise.

If the function we're supposed to be wrapping is of the form `fn(...args, (error, result) => {})` then that should probably be mentioned somewhere in the instructions, perhaps with examples of before/after, etc.

### PR DESCRIPTION
**Pull request by TomPradat**
Fix #1331 

TODO :
- [x] Make the exercise available again (by removing the wip status)

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -975,7 +975,7 @@
         "core": false,
         "practices": [],
         "prerequisites": [],
-        "difficulty": 4,
+        "difficulty": 6,
```
Comment (TomPradat) [inline config.json:978]: IMO the difficulty is underrated

File: exercises/practice/promises/.docs/introduction.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Introduction
```
Comment (TomPradat) [inline exercises/practice/promises/.docs/introduction.md:1]: I'm not sure how this file is displayed on the website, will it be visible by default before the instructions?

File: exercises/practice/promises/.docs/instructions.md [hunk 3]
```diff
@@ -1,11 +1,30 @@
...
+const fetchProductAsPromise = promisify(fetchProduct);
+
+// Now you got a function `fetchProductAsPromise`
+// that returns a promise
+fetchProductAsPromise(productId)
+    .then((data) => {})
+    .catch((error) => {});
+```
+
+2. Re-implement the following built-ins `Promise` methods (without using them)
```
Comment (TomPradat) [inline exercises/practice/promises/.docs/instructions.md:25]: Should we update the tests to make sure the pupils are not using the built-in methods?

Comment (junedev) [reply exercises/practice/promises/.docs/instructions.md:25]: We usually don't focus on testing that a student does not cheat so I don't think this should be a priority right now.

File: exercises/practice/promises/.docs/introduction.md (hunk 2 above)
Comment (junedev) [reply exercises/practice/promises/.docs/introduction.md:1]: I am pretty sure it will look the same as all the concept exercises where you first see the introduction text and then the instructions.

**Comment by junedev** [comment]
@joshgoebel Can you have a look whether you think the exercise is ok now?

File: exercises/practice/promises/.docs/introduction.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+Before the `Promise` class was introduced, there was only one way to deal with asynchronous code : the _callback pattern_.
+
+A callback is a function that accept an "error" as first parameter (see example below)
```
Comment (junedev) [inline exercises/practice/promises/.docs/introduction.md:5]: I think a function with a different pattern can still be a callback, so maybe rephrase this, maybe like this:
```suggestion
A callback is a function that is passed as an argument to another function and will be called once some action in that other function has finished. A common pattern for those callback functions is that they accept an "error" as first parameter (see example below).
```

File: exercises/practice/promises/.docs/introduction.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+function callback(error, arg2, arg3) {}
+```
+
+Now is it related to asynchronous code ?
+
+Historically, callbacks have been used in order to allow people to do some work after an asynchronous task was done and without blocking the whole program.
+
+```javascript
+fetchProduct(productId, function (error, data) {
+  if (error !== null) {
```
Comment (junedev) [inline exercises/practice/promises/.docs/introduction.md:17]: I am no expert on this pattern but shouldn't this be `if (error)` so it also works if error is `undefined` for example?

**Review by junedev — APPROVED** [review]
Left some minor comments.  
In general I would be ok to up-wip this and see whether students are ok with the exercise now.

File: exercises/practice/promises/.docs/introduction.md (hunk 5 above)
Comment (TomPradat) [reply exercises/practice/promises/.docs/introduction.md]: Good catch, fixed in 614fa7060fd3cb084c481a90fef2ed0772c798bb

**Comment by joshgoebel** [comment]
Let me pull this down and have a look.

**Comment by joshgoebel** [comment]
Much much better I think.  

```
export const all = (promises) =>
  promises.reduce(
    async (acc, promise) => (await acc).concat(await promise),
    Promise.resolve([])
  );
```

Just out of curiosity this isn't concurrent, is it?  It seems like each promise stacks, having to wait until the one before it finishes before the array is resolved and then concat for the next item can be called, correct?

**Review by junedev — APPROVED** [review]

**Comment by TomPradat** [comment]
> Much much better I think.
> 
> ```
> export const all = (promises) =>
>   promises.reduce(
>     async (acc, promise) => (await acc).concat(await promise),
>     Promise.resolve([])
>   );
> ```
> 
> Just out of curiosity this isn't concurrent, is it? It seems like each promise stacks, having to wait until the one before it finishes before the array is resolved and then concat for the next item can be called, correct?

I'm not used to using `reduce` with async functions but I have tested this in the browser and it looks like it is **CONCURRENT** :

```javascript
let all = (promises) =>
  promises.reduce(
    async (acc, promise) => (await acc).concat(await promise),
    Promise.resolve([])
  );
  
  let createPromise = () =>
      new Promise((resolve) => {
          setTimeout(() => {
               console.log(performance.now(), "hello");
               resolve();
          }, 1000);
      });

all([createPromise(), createPromise(), createPromise(), createPromise()]).then(() => console.log("finished"));
```

Result :

```
2540591.3000000003 'hello'
2540591.5000000005 'hello'
2540591.6 'hello'
2540591.6999999997 'hello'
finished
```

I must admit I don't quite understand why though (or maybe something is wrong with my test code).

**Comment by TomPradat** [comment]
@junedev Should i remove the "wip" status before this gets merged ?

**Comment by junedev** [comment]
yes please

**Comment by joshgoebel** [comment]
@TomPradat Are you sure?  You're creating them all at once (in the array creating)... so their setTimeout will be concurrent, that doesn't mean `reduce` is... I don't see how it can be... it has to wait, one at a time - since array addition is sequential.  Perhaps this doesn't matter in practice though since the time spent in reduce is so small...?

**Comment by TomPradat** [comment]
> @TomPradat Are you sure? You're creating them all at once (in the array creating)... so their setTimeout will be concurrent, that doesn't mean `reduce` is... I don't see how it can be... it has to wait, one at a time - since array addition is sequential. Perhaps this doesn't matter in practice though since the time spent in reduce is so small...?

You are perfectly right, I came to the same conclusion after some more thoughts. The `reduce` is sequential :

```javascript
let all = (promises) =>
  promises.reduce(
    async (acc, promise) => (await acc).concat(await promise()),
    Promise.resolve([])
  );
  
  let createPromise = () =>
      new Promise((resolve) => {
          setTimeout(() => {
               console.log(performance.now(), "hello");
               resolve();
          }, 1000);
      });

all([createPromise, createPromise, createPromise, createPromise]).then(() => console.log("finished"));
```
> Note : The difference with the previous sample is that the promise is being created in the `reduce` instead of being created in the array

Result :

```
3360643.1999999997 'hello'
3361644.1 'hello'
3362644.6999999997 'hello'
3363645.4 'hello'
finished
```

So to conclude, the `reduce` is sequential but it doesn't matter as the promises run in parallel. It adds a small overhead that is probably (IMO) negligible

**Comment by TomPradat** [comment]
> yes please

Done in ddd27138
